resources/2024-10-dapr-bindings.md

```diff
@@ -0,0 +1,368 @@
+
+This implementation will also have some side-effects outside of the Applications.Dapr resource provider:
+- Updates to the allowed resource type in the Portable Resource Renderer
+- Updates to the `getResourceDataByID` function in the Core RP to take the new type into account 
+- Updates to the `ResourceTypesList` variable in the CLI (cli/clients/managements.go) to be able to list the new type
+
+### Error Handling
+
+For input bindings, the Dapr sidecar will send an OPTIONS request to the application at startup. The application must respond with either a 2xx or a 405 status code. If the application does not respond, the Dapr sidecar will retry the initialization 3 times before canceling the binding initialization. Additional checks can also be configured on a per-binding basis.
+
+For output bindings, if there is an error with the external system, the Dapr sidecar will [retry the operation with the default policy](https://docs.dapr.io/operations/resiliency/policies/#retries).
+
+This [behavior is configurable](https://docs.dapr.io/operations/resiliency/policies/#overriding-default-retries) by the user, though it is not yet possible to configure the [resiliency policy](https://docs.dapr.io/operations/resiliency/resiliency-overview/#complete-example-policy) in Radius.
+
+
+## Test plan
+
+Unit tests must cover:
+- converter functions
+- processor functions
+
+Functional tests proposed:
+- Using a binding in a sample application (the direction, input/output, of the binding is not important for this test). The external system can be something simple like Redis or a [CRON binding](https://docs.dapr.io/reference/components-reference/supported-bindings/cron/) that doesn't require an external system.
+- Using a binding with secret-store indirection, which will require creating and using a secret store in the binding.
+
+## Security
+
+For data at rest, the binding can reference a secret store instead of storing sensitive information directly within the binding configuration.
+
+For data in transit, all communication between Dapr components and external systems should use TLS/SSL. This can be configured on a per-binding basis in the metadata. 
+
+Additionally, networking configurations can be applied in accordance with the [Dapr security specification](https://docs.dapr.io/concepts/security-concept/#bindings-security).
+
+
+
+## Compatibility (optional)
+
+This is purely additive and should not affect existing resources.
+
+## Monitoring and Logging
+
+N/A
+
+## Development plan
+
+N/A, already completed 
+
+## Open Questions
+N/A
+
+## Alternatives considered
+
+N/A 
+
+## Design Review Notes
+
+Round 1:
+- In the first iteration, recipes were considered out of scope, as they would result in a two-step process to create a single CRD (the Dapr Binding component). This was reconsidered, as not supporting recipes for a resource contradicts the Radius design philosophy. Therefore, common usage recipes were added to the design.
```

Comment:
I don't quite understand this - apologies, I would out of the office when this was discussed. We have existing recipes for Dapr components that create both the infrastructure and the Dapr `Component` resource, so I'm not aware of any issues. LMK if I'm missing anything!

Reply:
Sorry for the delay!

There is no issue with the creation process itself. You can create both the Binding CRD and the infrastructure just as you would with any other Dapr Building Block. Users can (and likely will) do this with their own configurations, and we certainly shouldn’t try to restrict that flexibility.

My point is only relevant to the initial binding recipes we plan to provide to users. Here’s some context: the Bindings Building Block is unique in Dapr. It predates most other building blocks and originally served as a way to accomplish tasks that did not yet have dedicated building blocks. This is why some bindings overlap with other blocks, such as Azure Storage Queue.

Bindings are generally basic, often supporting only CRUD-like operations. Additionally, Bindings is the only building block without a unified feature set across multiple backends—nearly every binding is different, and few are interchangeable.

Today, bindings are intended for interacting with external resources, as they are designed for basic interactions. For internal, app-scoped backing services, there is usually a better building block available.

While this approach doesn’t affect how Radius implements Bindings, it does impact the default recipes we include in radius/recipes. If we consider bindings best suited for external resources, then a binding recipe should only contain the Dapr Component CRD. This is why I initially opposed creating binding recipes, as it would involve a two-step process for what is effectively a single CRD creation.

In the first meeting, however, a very good point was raised: from Radius's perspective, it is difficult to justify not having any recipes for a resource.

So, my revised proposal is to create default recipes with this "external resource" perspective in mind. These recipes would set up the backing service in local-dev for convenience, but in production, they would have as input the credentials to an existing backing service. This approach isn’t meant to restrict users but rather as a way to nudge users in the right direction from the start. Of course, users can create custom recipes to override this behavior as they see fit.

I'm not entirely sure if I explained this clearly or if it make sense, so please don’t hesitate to challenge me or ask for clarification on any part of this! And if this direction makes sense, I’m happy to revise the design doc to reflect it.

Reply:
> So, my revised proposal is to create default recipes with this "external resource" perspective in mind. These recipes would set up the backing service in local-dev for convenience, but in production, they would have as input the credentials to an existing backing service. This approach isn’t meant to restrict users but rather as a way to nudge users in the right direction from the start. Of course, users can create custom recipes to override this behavior as they see fit.

Hey @SoTrx - I appreciate the response 👍 

I think the part where I'm stuck is why we need **recipes** for the "pre-existing" case. This is what `resourceProvisioning: 'manual'` is for.... when you want to configure the metadata directly.  

I'm happy to continue this discussion when we start looking at recipe examples since it's not really going to change the design of the `bindings` resource type.